src/main/resources/include/hail/Region.h

```diff
@@ -0,0 +1,176 @@
+    long  size_;
+    
+    inline BigAlloc() : buf_(nullptr), size_(0) { }
+
+    inline BigAlloc(char* buf, long size) : buf_(buf), size_(size) { }
+    
+    inline BigAlloc(const BigAlloc& b) : buf_(b.buf_), size_(b.size_) { }
+    
+    inline BigAlloc& operator=(const BigAlloc& b) {
+      buf_ = b.buf_;
+      size_ = b.size_;
+      return(*this);
+    }
+  };
+
+public:
+  char* chunk_;
+  long  pos_;
+  std::vector<char*> free_chunks_;
+  std::vector<char*> full_chunks_;
+  std::vector<BigAlloc> big_free_;
+  std::vector<BigAlloc> big_used_;
+  std::vector<RegionPtr> required_regions_;
+
+public:  
+  Region() :
+    chunk_(nullptr),
+    pos_(kChunkCap) {
+  }
+  
+  virtual ~Region() {
+    if (chunk_) free(chunk_);
+    for (auto p : free_chunks_) free(p);
+    for (auto p : full_chunks_) free(p);
+    for (auto& b : big_free_) free(b.buf_);
+    for (auto& b : big_used_) free(b.buf_);
+  }
+  
+  // clear_but_keep_mem() will make the Region empty without free'ing chunks
+  void clear_but_keep_mem() {
+    pos_ = (chunk_ ? 0 : kChunkCap);
+    for (auto p : full_chunks_) free_chunks_.push_back(p);
+    full_chunks_.clear();
+    // Move big_used_ to big_free_
+    for (auto& used : big_used_) {
+      big_free_.emplace_back(used.buf_, used.size_);
+      used.buf_ = nullptr;
+    }
+    big_used_.clear();
+    // Sort in descending order of size
+    std::sort(
+      big_free_.begin(),
+      big_free_.end(),
+      [](const BigAlloc& a, const BigAlloc& b)->bool { return(a.size_ > b.size_); }
+    );
+    if (big_free_.size() > kNumBigToKeep) {
+      for (int idx = big_free_.size(); --idx >= (int)kNumBigToKeep;) {
+        free(big_free_[idx].buf_);
+      }
+      big_free_.resize(kNumBigToKeep);
+    }
+  }
+  
+  void new_chunk() {
+    if (chunk_) full_chunks_.push_back(chunk_);
+    if (free_chunks_.empty()) {
+      chunk_ = (char*)malloc(kChunkCap);
+    } else {
+      chunk_ = free_chunks_.back();
+      free_chunks_.pop_back();
+    }
+    pos_ = 0;
+  }
+  
+  // Restrict the choice of block sizes to improve re-use
+  long choose_big_size(long n) {
+    for (long b = 1024;; b <<= 1) {
+      if (n <= b) return b;
+      // sqrt(2) is 1.414213, 181/128 is 1.414062
+      long bmid = (((181*b) >> 7) + 0x3f) & ~0x3f;
+      if (n <= bmid) return bmid;
+    }
+  }
+  
+  char* allocate_big(long n) {
+    char* buf = nullptr;
+    //
+    n = ((n + 0x3f) & ~0x3f); // round up to multiple of 64byte cache line
+    for (int idx = big_free_.size(); --idx >= 0;) {
+      auto& b = big_free_[idx];
+      if (n <= big_free_[idx].size_) {
+        buf = b.buf_;
+        b.buf_ = nullptr;
+        big_used_.emplace_back(buf, b.size_);
+        // Fast enough for small kNumBigToKeep
+        big_free_.erase(big_free_.begin()+idx);
+        return(buf);
+      }
+    }
+    n = choose_big_size(n);
+    buf = (char*)malloc(n);
+    big_used_.emplace_back(buf, n);
+    return(buf);
+  }
+  
+  inline void align(long a) {
```

Comment:
I still like it.  In the design process, I'm thinking about which methods should *always* be inlined for performance reasons, and I want to document that intent with an explicit "inline" even if the compiler then chooses to ignore it.

If the Google Style Guide recommends otherwise, I'll fit in with it.  But I'm not keen to try
to follow a whole lot of rules from other places.

Reply:
Google only says "Do not put large method definitions inline in the class definition" and gives rules of thumb for when to declare functions inline. They're silent on this question.

To be sure we're on the same page: I'm not saying anything about whether the compiler chooses to ignore the `inline` annotation. I'm saying that functions defined inside a class body are `inline` whether you write it or not. You have some larger functions defined in the class body without `inline` annotations, but they are also considered `inline` by the compiler (which of course it is free to ignore).

It's a very minor and purely syntactic thing, I would just like to be consistent whatever we decide. @cseed do you have a preference?

Reply:
It gets fairly confusing.  We can say for sure that the compiler will *not* inline a function
at the point of a call if a) the call is in a separate compilation unit from the function definition,
and b) it isn't attempting some wacky whole-program optimization (which AFAIK no-one
really does outside of artificial benchmarks).

Now if the method is inside the class declaration, then any compilation unit can see it and the compiler is capable of inlining it.  But it may choose not to do so, depending on the context,
and how big the method is, and the compiler options, and the phase of the moon.

[Aside: the ambiguity about when inlining does and doesn't happen is another reason
for preferring macros - which are definitely always inlined, no questions asked - under some
circumstances.  Especially for dealing with stuff where compiling in debug mode might
inhibit inlining and cause a huge slowdown ...]

This is another issue where the mixed Scala/C++ environment changes the picture slightly -
in Java/Scala the language encourages us to write everything inside the class declaration and let the compiler and JVM figure out how to optimize it.  And if that style bleeds over into our C++, then we'll probably have quite a lot of relatively large implicitly-inline methods.  Which I think is mostly ok (though we shouldn't expose many large methods in header files), but means that methods that "really-really-want-to-be-inline" have to be distinguished in
some other way.

Reply:
My rules are:

1. If it's small enough that I definitely always want it to be inline'd, then I put it in the
    class body with an explicit "inline".

2. If it seems more readable to have the body of the function in the class body, I also
    put it in the class body, but without the explicit "inline".

3. If it's big enough that I think a reasonable human would prefer to read the class body
    and only go searching for the method body if they really need to look at it, then I
    write it separately (which adds a few extra lines of code).

4. If it's a class declared in a header file, then the balance swings more towards hiding
   the method implementations in the .cpp

Over time, I tend to have been evolving towards using case 2 more frequently, since that
makes it look more like the other commonly-used languages (Java, Scala, Python etc) and
so is arguably easier to read for people who don't have a lot of C++ experience.

I'm basically in favor of writing C++ which conforms to the semantic and syntactic style
of more modern languages where it doesn't make much difference (and in terms of
performance and memory usage, usually only 5-10% of the source code needs to be
carefully optimized, and the other 95% just needs to be correct and readable), and uses the 
features and style peculiar to C++ only where those give a substantial win.  Hence my attitude to
std::unique_ptr and move-semantics.